text/0000-default-field-values.md

```diff
@@ -0,0 +1,1767 @@
+[motivation]: #motivation
+
+## Boilerplate reduction
+
+### For `struct`s
+
+[FRU]: https://doc.rust-lang.org/1.31.0/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax
+
+Rust allows you to create an instance of a `struct` using the struct literal
+syntax `Foo { bar: expr, baz: expr }`. To do so, all fields in the `struct`
+must be assigned a value. This makes it inconvenient to create large `struct`s
+whose fields usually receive the same values. Struct literals also cannot be
+used to initialize `struct`s with fields which are inaccessible due to privacy.
+*[Functional record updates (FRU)][FRU]* can reduce noise when a `struct`
+derives `Default`, but are also invalid when the `struct` has inaccessible
+fields and do not allow the creation of an `impl` where *some* fields are
+mandatory.
+
+To work around these shortcomings, you can create constructor functions:
+
+```rust
+struct Foo {
+    alpha: &'static str,
+    beta: bool,
+    gamma: i32,
+}
+
+impl Foo {
+    /// Constructs a `Foo`.
+    fn new(alpha: &'static str, gamma: i32) -> Self {
+        Self {
+            alpha,
+            beta: true,
+            gamma
+        }
+    }
+}
+
+let foo = Foo::new("Hello", 42);
+```
+
+[`process::Command`]: https://doc.rust-lang.org/stable/std/process/struct.Command.html
+
+The problem with a constructor is that you need one for each combination
+of fields a caller can supply. To work around this, you can use builders,
+such as [`process::Command`] in the standard library.
+Builders enable more advanced initialization, but require additional boilerplate.
+To represent the difference, we can see the dramatic syntactical increase for
+semantically small changes:
+
+```rust
+// All fields are mandatory
+struct Foo {
+    alpha: &'static str,
+    beta: bool,
+    gamma: i32,
+}
+```
+
+```rust
+impl Foo {
+    /// Constructs a `Foo`.
+    fn new(alpha: &'static str, gamma: i32) -> Self {
+        Self {
+            alpha,
+            beta: true,
+            gamma
+        }
+    }
+}
+```
+
+```rust
+// A builder type that is able to construct a `Foo`, but that will fail at runtime if a field is
+// missing.
+#[derive(Default)]
+struct FooBuilder {
+    pub alpha: Option<&'static str>,
+    pub beta: Option<bool>,
+    pub gamma: Option<i32>,
+}
+
+impl FooBuilder {
+    fn new() -> Self {
+        FooBuilder::default()
+    }
+    fn set_alpha(&mut self, alpha: &'static str) -> &mut Self {
+        self.alpha = Some(alpha);
+        self
+    }
+    fn set_beta(&mut self, beta: &'static str) -> &mut Self {
+        self.beta = Some(beta);
+        self
+    }
+    fn set_gamma(&mut self, gamma: &'static str) -> &mut Self {
+        self.gamma = Some(gamma);
+        self
+    }
+
+    fn build(self) -> Foo {
+        Foo {
+            alpha: self.alpha.unwrap(),
+            beta: self.beta.unwrap(),
+            gamma: self.gamma.unwrap_or(0),
+        }
+    }
+}
+```
+
+```rust
```

Comment:
I think it's worth also comparing against derive macros like [typed-builder](https://github.com/idanarye/rust-typed-builder), which at least in my experience are a very common solution to this problem.